[RunAllTests] Fix #3451: [RTL] High-fi Add support for RTL in binding adapters

veena14cs opened the pull request:
## Explanation
This PR fixes #3451. This PR fixes the custom attributes used in data binding adapters to set margin or padding.
This fix was mainly for `content` and `feedback` items of Exploration player.

This PR also fixes #3205.

## Screenshot LTR and RTL
Mobile:
<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125128607-d80c8d80-e11b-11eb-8eae-8cfbd7e59703.png">......<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125128772-1f931980-e11c-11eb-9b9c-8e661c65b8df.png">

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125128638-e6f34000-e11b-11eb-800f-6427e4e1de85.png">...... <img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125128682-fc686a00-e11b-11eb-9728-d842b2eba3f2.png">



Tablet:
<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125129400-09398d80-e11d-11eb-9f3b-2bd8cacf7020.png">....<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125129547-4140d080-e11d-11eb-868a-ae5796f16368.png">



<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125129438-18b8d680-e11d-11eb-806f-ec979d4e3672.png">.....<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125129502-2d956a00-e11d-11eb-923f-b279f60bfc1d.png">


## Expresso Tests
![Screenshot (20)](https://user-images.githubusercontent.com/16301028/126455352-2257ad2e-4ba0-4488-8493-5d0a6436a27b.png)


![Screenshot (21)](https://user-images.githubusercontent.com/16301028/126455364-285d6e27-8fe1-49d3-a0a4-bdc4c9b3e5c3.png)


![Screenshot (23)](https://user-images.githubusercontent.com/16301028/126457727-657b09e5-163b-4a85-89c3-59d0c4f6e8ac.png)


## Screenshots of tests that fails as expected when RTL code is temporarily removed.


![Screenshot (32)](https://user-images.githubusercontent.com/16301028/126841773-e0b16d4e-dc66-4a34-9004-f4b3e7d8b634.png)

![Screenshot (30)](https://user-images.githubusercontent.com/16301028/126841954-fb347a00-cddd-447f-8a6d-9ae82a5b82b8.png)


![Screenshot (34)](https://user-images.githubusercontent.com/16301028/126841800-a571cb54-069c-4794-a8bc-e9291cd19cd0.png)



## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Added check if RTL is enabled

veena14cs commented:
This PR also fixes #3205. Once this is verified and approved will close PR #3330 

## Screenshot LTR and RTL

Mobile

<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271569-2a7fd300-cefd-11eb-8733-89198f6dd44f.png">  .... <img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125134121-8c121680-e124-11eb-8f63-d45a90992423.png">

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122271508-19cf5d00-cefd-11eb-9a8e-41967a5ef2b1.png">  ... <img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125134100-7ef52780-e124-11eb-96d1-e4cf777ffdf6.png">


Tablet

<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/122339806-2389ac80-cf5f-11eb-94af-4d6a690a81de.png">.....<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125133918-2c1b7000-e124-11eb-9cc9-5602e8093f6b.png">
<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/124327096-30381280-dba5-11eb-83e8-1ad1d642f422.png">.......<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125133971-3c334f80-e124-11eb-8ea4-300c837ef32e.png">

rt4914 commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,22 +1,35 @@
+        float temp = marginStart;
+        marginStart = marginEnd;
+         marginEnd = temp;
+      }
```
Formatting updates:
```
      if (isRtlLayout(view)) {
        float temp = marginStart;
        marginStart = marginEnd;
        marginEnd = temp;
      }
```

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -28,8 +41,14 @@ public static void setLayoutMarginStart(@NonNull View view, float marginStart) {
     if (view.getLayoutParams() instanceof MarginLayoutParams) {
       MarginLayoutParams params = (MarginLayoutParams) view.getLayoutParams();
+      float marginStart = params.getMarginStart();
+      if(isRtlLayout(view)){
```
Nit spacing
`if (isRtlLayout(view)) {`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdapters.java:
```diff
@@ -106,6 +113,11 @@ private static void setMarginIfApplicable(
       float marginEnd,
       float marginBottom
   ) {
+    if(isRtlLayout(view)){
```
Nit spacing
`if (isRtlLayout(view)) {`

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdapters.java:
```diff
@@ -105,6 +111,11 @@ private static void setPaddingIfApplicable(
       float paddingEnd,
       float paddingBottom
   ) {
+    if(isRtlLayout(view)){
```
Nit spacing
`if (isRtlLayout(view)) {`

veena14cs replied:
Done,

rt4914 commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,22 +1,35 @@
 
+  @SuppressLint("WrongConstant")
+  private static boolean isRtlLayout(View view) {
+    return view.getLayoutDirection() == ViewCompat.LAYOUT_DIRECTION_RTL;
```
@veena14cs Has this been confirmed by @BenHenning ? The use of `LAYOUT_DIRECTION_RTL ` and also `SuppressLint` ?

veena14cs replied:
@BenHenning can you please look into this.

BenHenning replied:
We should be using https://developer.android.com/reference/androidx/core/view/ViewCompat#getLayoutDirection(android.view.View) since that will work on older Android versions.

I suspect that will also fix the lint issue and that we can then remove the suppression.

This approach seems reasonable given that it's the only clear way to check whether the current layout is RTL.

veena14cs replied:
Thanks. Done this change.

rt4914 replied:
Thanks.

rt4914 reviewed (changes requested):
@veena14cs PTAL, thanks.

veena14cs committed:
Fixed spacing

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,35 +1,61 @@
 
-  /** Used to set a margin-start for views. */
+  @SuppressLint("WrongConstant")
+  private static boolean isRtlLayout(View view) {
```
Nit: please move to be below the ``public`` methods.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,35 +1,61 @@
 import androidx.databinding.BindingAdapter;
 
-/** Holds all custom binding adapters that set margin values. */
+/**
+ * Holds all custom binding adapters that set margin values.
+ */
```
Here & below: prefer single-line Javadocs if they fit, e.g.:

```java
/** Custom databinding adapters that set margin values. */
```

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,35 +1,61 @@
+  }
+
+  /**
+   * Used to set a margin-start for views.
```
Here & elsewhere: method Javadocs should be a singular third-person verb or verb phrase. To adjust this one to match:

```java
/** Sets the start margin for a view, accounting for RTL scenarios. */
```

(Plus the additional RTL bit that's being added here).

Ditto elsewhere.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -1,35 +1,61 @@
+        marginEnd = temp;
+      }
       params.setMargins(
           (int) marginStart,
           params.topMargin,
-          params.getMarginEnd(),
+          (int) marginEnd,
           params.bottomMargin
       );
```
Maybe consider instead creating a helper method to set margins which automatically takes into consideration RTL. That will avoid needing an if-check in each binding method since they'd share a common margin value setter.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdapters.java:
```diff
@@ -106,6 +113,11 @@ private static void setMarginIfApplicable(
       float marginEnd,
       float marginBottom
   ) {
+    if (isRtlLayout(view)) {
```
For here & the previous file, I suggest introducing marginLeft & marginRight values that are passed into setMargin (since those are the actual values being set), and to initialize them in an if/else based on RTL. That will actually be a lot more clear on what's happening here since we aren't actually switching the start/end, we're properly deriving left/right based on start/end (which is affected by RTL).

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdapters.java:
```diff
@@ -105,6 +111,11 @@ private static void setPaddingIfApplicable(
       float paddingEnd,
       float paddingBottom
   ) {
+    if (isRtlLayout(view)) {
```
Ditto here (but for paddingLeft/paddingRight).

veena14cs replied:
Done,

BenHenning reviewed (changes requested):
Thanks @veena14cs. Left some thoughts. Could you please also add tests for each of these adapters classes? We should make sure that the new RTL logic correctly sets left/right margins & padding when expected.

veena14cs committed:
fixed nit issues

veena14cs committed:
optimized code

veena14cs commented:
> Thanks @veena14cs. Left some thoughts. Could you please also add tests for each of these adapters classes? We should make sure that the new RTL logic correctly sets left/right margins & padding when expected.

@BenHenning Do you have any example on how we can test these binding adapters. I am finding it difficult.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -8,33 +8,52 @@
+    }
+  }
+
+  private static void setLayoutDirectionalMargins(
```
Some nits:
1. Indentation for these should be 2 spaces
2. We can we-retrieve the params rather than needing to pass them in again (since we have the view, anyway)
3. marginStart should be before marginEnd (as is conventional in Android)

veena14cs replied:
Done.

BenHenning replied:
Point (1) still seems unaddressed. Ditto elsewhere in PR (we should use +2 indentation for Java in these cases).

veena14cs replied:
Done.

BenHenning replied:
This still seems unresolved. This is how the indentation should be:

```kotlin
  private static void setLayoutDirectionalMargins(
    @NonNull View view,
    int marginStart,
    int marginEnd
  ) {
    MarginLayoutParams params = (MarginLayoutParams) view.getLayoutParams();
    params.setMargins(
      marginStart,
      params.topMargin,
      marginEnd,
      params.bottomMargin
    );
  }
```

Ditto elsewhere.

Also: please don't mention something is 'done' without the changes being applied. If you're unsure what's being asked, please ask for clarification.

veena14cs replied:
This is already fixed.

BenHenning replied:
No, it isn't. This is the current version on GitHub: https://github.com/oppia/oppia-android/blob/2c454f307dd23a650bbff9071488e1b194e99ecc/app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java#L44

It has 4 spaces of indentation, not 2. If this is fixed for you locally, are you pushing it?

BenHenning replied:
It appears fixed now. Thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -8,33 +8,52 @@
 public final class MarginBindingAdapters {
 
-  /** Used to set a margin-start for views. */
+  public static boolean isRtlLayout(View view) {
```
This should be private since it's not a binding adapter. I also suggest moving it to be the last method in the class since it's a helper (and we generally organize classes "appendix style" where helpers appear after the place where theyr'e needed).

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdapters.java:
```diff
@@ -10,6 +10,10 @@
  */
 public final class StateAssemblerMarginBindingAdapters {
 
+  public static boolean isRtlLayout(View view) {
```
Ditto here.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdapters.java:
```diff
@@ -9,6 +9,9 @@
  */
 public final class StateAssemblerPaddingBindingAdapters {
 
+  public static boolean isRtlLayout(View view) {
```
Ditto here.

veena14cs replied:
Done.

BenHenning reviewed (changes requested):
Thanks @veena14cs. Left some more thoughts while the tests are being added.

I provided you some examples over chat. Please let me know (over chat) if anything else is needed.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs commented:
> Thanks @veena14cs. Left some more thoughts while the tests are being added.
> 
> I provided you some examples over chat. Please let me know (over chat) if anything else is needed.

Thanks @BenHenning for example  and @anandwana001 for your PR #3219, that really helped me to write testcases. @anandwana001 can you please look into the testcases in this PR.

veena14cs committed:
Added testcases

veena14cs committed:
removed unused imports

BenHenning commented:
I may need to look at this tomorrow, but could you please fix the failing tests being reported on CI in the meantime @veena14cs?

veena14cs committed:
corrected spacing

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update StateAssemblerPaddingBindingAdapters.java

veena14cs committed:
updated scripts

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
fixed spacing issue

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update StateAssemblerPaddingBindingAdaptersTest.kt

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -216,6 +218,7 @@
       android:name=".app.devoptions.vieweventlogs.ViewEventLogsActivity"
       android:label="@string/view_event_logs_activity_title"
       android:theme="@style/OppiaThemeWithoutActionBar" />
+    <activity android:name=".app.testing.MarginBindableAdaptersTestActivity" />
```
Please add this in alphabetical order.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/activity/ActivityComponent.kt:
```diff
@@ -122,6 +126,14 @@ interface ActivityComponent {
   fun inject(splashActivity: SplashActivity)
   fun inject(splashTestActivity: SplashTestActivity)
   fun inject(stateFragmentTestActivity: StateFragmentTestActivity)
+  fun inject(
```
Please inject these in alphabetical order.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/MarginBindableAdaptersTestActivity.kt:
```diff
@@ -0,0 +1,14 @@
+import org.oppia.android.R
+import org.oppia.android.app.activity.InjectableAppCompatActivity
+
+/** This is a dummy activity to test MarginBindableAdapters. */
```
/** This is a test activity for MarginBindableAdapters. */

BenHenning replied:
Suggest just: ``Test activity for MarginBindableAdapters.`` per Rajat's comment (the bits before 'Test activity' are unnecessary in-context). Ditto elsewhere in this PR.

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/StateAssemblerMarginBindingAdaptersTestActivity.kt:
```diff
@@ -0,0 +1,14 @@
+import org.oppia.android.R
+import org.oppia.android.app.activity.InjectableAppCompatActivity
+
+/** This is a dummy activity to test StateAssemblerMarginBindingAdapters. */
```
Ditto as above

veena14cs replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/android/app/testing/StateAssemblerPaddingBindingAdaptersTestActivity.kt:
```diff
@@ -0,0 +1,14 @@
+import org.oppia.android.R
+import org.oppia.android.app.activity.InjectableAppCompatActivity
+
+/** This is a dummy activity to test StateAssemblerPaddingBindingAdapters . */
```
Ditto as above

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
PTAL @veena14cs 

Also, please add screenshots displaying that tests are passing on Espresso PR Description

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -82,4 +96,8 @@ public static void setLayoutMargin(@NonNull View view, float margin) {
   }
+
+  private static boolean isRtlLayout(View view) {
+    return view.getLayoutDirection() == View.LAYOUT_DIRECTION_RTL;
```
Per my earlier comment, shouldn't we be using ViewCompat to get the layout direction, instead?

Ditto in other files in this PR.

veena14cs replied:
Earlier it was using ViewCompat, due to which it has lint issues. I changed it to view and it fixed lint issues and removed suppression as per your comment  https://github.com/oppia/oppia-android/pull/3452#discussion_r668043999

veena14cs replied:
Earlier I was using ViewCompat and it was giving lint issues so changed it to View.

BenHenning replied:
I think your lint issues were because you weren't using the constants from ViewCompat itself (such as https://developer.android.com/reference/androidx/core/view/ViewCompat#LAYOUT_DIRECTION_RTL).

BenHenning commented on app/src/main/res/layout/test_margin_bindable_adapter_activity.xml:
```diff
@@ -0,0 +1,20 @@
+    android:layout_height="wrap_content">
+
+    <TextView
+      android:id="@+id/test_margin_textview"
```
```suggestion
      android:id="@+id/test_margin_text_view"
```

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,363 @@
+      return@runWithActivity textView
+    }
+    assertThat(textView.marginStart.toFloat()).isEqualTo(
+      context.resources.getDimension(R.dimen.content_item_exploration_view_margin_start)
```
I suggest creating helpers to retrieve the start/end dimensions here & in the other tests to simplifying the tests themselves.

veena14cs replied:
done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,363 @@
+
+  @Config(qualifiers = "sw600dp-port")
+  @Test
+  fun testMarginBindableAdapters_ltrEnabled__port_tablet_marginStartAndMarginEndForLtrIsCorrect() {
```
```suggestion
  fun testMarginBindableAdapters_ltrEnabled_port_tablet_marginStartAndMarginEndForLtrIsCorrect() {
```

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,363 @@
+      )
+      return@runWithActivity textView
+    }
+    assertThat(textView.marginStart.toFloat()).isEqualTo(
```
We should either convert to integers, or use isWithin for proper float comparison:

```kotlin
assertThat(floatSubject).isWithin(TOLERANCE).of(expectedValue)
```

Where TOLERANCE is something like:

```kotlin
private const val TOLERANCE = 1e-5f
```

veena14cs replied:
Used `assertThat(floatSubject).isWithin(TOLERANCE).of(expectedValue)` for float comparison.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,363 @@
+  fun testMarginBindableAdapters_rtlIsEnabled_landscape_marginStartAndMarginEndForRtlIsCorrect() {
+    val textView = activityRule.scenario.runWithActivity {
+      val textView: TextView = it.findViewById(R.id.test_margin_textview)
+      setLayoutMarginStart(
+        textView,
+        context.resources.getDimension(R.dimen.content_item_exploration_view_margin_start)
+      )
+      setLayoutMarginEnd(
+        textView,
+        context.resources.getDimension(R.dimen.content_item_exploration_view_margin_end)
+      )
```
Could this bit be moved to a helper to simplify the tests? Ditto in other test files.

veena14cs replied:
Done here and to other test files too.

BenHenning reviewed (changes requested):
Thanks @veena14cs! Had some thoughts. I only looked at one of the test files for now since they're a bit redundant. After you address the open comments that also apply to those, I'll do another pass in the next round of reviews.

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs committed:
fixed nit changes

veena14cs committed:
Update app/src/main/res/layout/test_margin_bindable_adapter_activity.xml

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs committed:
added helper methods

veena14cs committed:
added helper methods

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs committed:
fixed issues

veena14cs committed:
Update StateAssemblerMarginBindingAdaptersTest.kt

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update ActivityComponent.kt

veena14cs committed:
Update ActivityComponent.kt

rt4914 reviewed (approved):
Approving for code-owner files. Thanks @veena14cs 

Also, do you think we can apply same thing for PinPasswordActivity now?

BenHenning commented on app/src/main/res/values-land/dimens.xml:
```diff
@@ -65,6 +65,7 @@
   <!-- Content Item: Exploration View -->
   <dimen name="content_item_exploration_view_margin_start">48dp</dimen>
   <dimen name="content_item_exploration_view_margin_top">24dp</dimen>
+  <dimen name="content_item_exploration_view_margin_bottom">24dp</dimen>
```
If this dimension is only needed for tests, could you instead create a new test-dimens.xml file for this to be part of? We should keep test-only artifacts separate from production ones.

(And ditto for the other values folders).

BenHenning replied:
Actually, please disregard this in favor of the suggestion below to hardcode the dimensions, instead.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+  }
+
+  private fun getContentMarginEnd(): Float {
+    return context.resources.getDimension(R.dimen.content_item_exploration_view_margin_end)
```
Instead of using resource-based dimensions, I think we should just hardcode the start/end values here. This may actually be preferable since it guarantees the test is using different start/end margin values.

Ditto for the other test suites.

veena14cs replied:
Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,455 @@
+
+  @Config(qualifiers = "sw600dp-land")
+  @Test
+  fun testMarginBindingAdapters_ltrEnabled_landscapeTablet_marginStartAndEndForLtrIsCorrect() {
```
```suggestion
  fun testMarginBindingAdapters_ltrEnabled_landscape_tablet_marginStartAndEndForLtrIsCorrect() {
```

(for consistency with your test name above). Please make sure all new tests in each of the new test suites are consistent in this way.

BenHenning reviewed (changes requested):
Thanks @veena14cs! Just had a few more comments.

Also, could you please add screenshots to your main PR comment showing each of the test suites failing when the new RTL code is temporarily removed from the adapter classes? We should make sure that each of the RTL tests will fail as expected.

veena14cs commented:
> Approving for code-owner files. Thanks @veena14cs
> 
> Also, do you think we can apply same thing for PinPasswordActivity now?

I am not sure, but I will test and let you know.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+/** Tests for [MarginBindableAdapters]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = MarginBindableAdaptersTest.TestApplication::class, qualifiers = "port-xxhdpi")
```
should we skip the passing qualifier here as all the tests had qualifier annotations?

veena14cs replied:
I tried doing this but it started me throwing Runtime errors. I couldn't dig it up in detail to find the error, so reverted the change.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+    return context.resources.getDimension(R.dimen.content_item_exploration_view_margin_start)
+  }
+
+  private inline fun <reified V, A : Activity> ActivityScenario<A>.runWithActivity(
```
Is it ok to generalize this function, so that we can use it from a central utility wherever it required and we can its own test file too.

veena14cs replied:
May be we can do that, we can have a separate PR to generalize this function otherwise.

anandwana001 replied:
Sounds good Let's create an issue for this.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,455 @@
+  }
+
+  @Test
+  fun testMarginBindingAdapters_explorationViewMargin_rtlIsEnabled_marginsForRtlIsCorrect() {
```
This test is a copy of the other 4 test cases, the other four test cases having annotations of portland and both for tablet port/land too, Is it necessary to have the same test cases 5 times? Any specific reason or can we have one test case and depend on it that it will pass on all the configuration as currently, it's also the same as all 5 test cases are same.

veena14cs replied:
Nice catch. But it has to test for "ExplorationViewMargin" which was missing. I have changed this test to `setExplorationViewMargin` and other tests are for setExplorationSplitViewMargin().

veena14cs replied:
Ah, no actually this case has to test for `setExplorationViewMargin` and not `setExplorationSplitViewMargin` I corrected this please check.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,455 @@
+  private fun setQuestionViewMargins(textView: TextView) {
+    setQuestionViewMargin(
+      textView,
+      /* isApplicable= */ true,
+      getExplorationMarginStart(),
+      getExplorationMarginTop(),
+      getExplorationMarginEnd(),
+      getExplorationMarginBottom()
```
```suggestion
      /* isApplicable= */ true,
      /* marginStart= */ getExplorationMarginStart(),
      /* marginTop= */ getExplorationMarginTop(),
      /* marginEnd= */ getExplorationMarginEnd(),
      /* marginBottom= */ getExplorationMarginBottom()
```

anandwana001 replied:
Looks clean I guess, feel free to skip if suggestion doesn't seems interesting

veena14cs replied:
Added inline comments, Thanks.

anandwana001 reviewed (changes requested):
Left few thoughts
Thanks @veena14cs

veena14cs commented:
> Also, could you please add screenshots to your main PR comment showing each of the test suites failing when the new RTL code is temporarily removed from the adapter classes? We should make sure that each of the RTL tests will fail as expected.

 I run the tests by temorarily removing RTL code, but all the tests are passing its not failing as expected. I suspect `textview.marginStart `is returning android's default margin attribute and not the Binding Adapter's attribute.

veena14cs committed:
used hardcoded dimens

veena14cs commented:
https://github.com/oppia/oppia-android/pull/3452#discussion_r674504808 This is already fixed. I  tried copy pasting and used ktlint for formatting. It doesn't show me any difference. Let me know if it is still having wrong indentation.

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  @Test
+  fun testMarginBindableAdapters_rtlIsEnabled_port_marginStartAndMarginEndForRtlFails() {
```
Test case where RTL start and end margin fails.

BenHenning replied:
Isn't this test backwards? We're asserting that the start/end are staying the same rather than flipping.

veena14cs replied:
This test I added as example to show how it is working. I will be removing this.

veena14cs committed:
fixed nit

veena14cs committed:
optimized code

BenHenning commented on app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:
```diff
@@ -97,7 +99,8 @@ public static void setLayoutMargin(@NonNull View view, float margin) {
     }
   }
 
+  @SuppressLint("WrongConstant")
```
Sorry, I think you may have misunderstood my previous comment. I think that this lint error is coming from the fact that we're mixing View.getLayoutDirection and ViewCompat. We need to use VIewCompat for both the layout direction retrieval and the constant.

veena14cs replied:
Yeah correct I misunderstood. Now I have fixed it.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -324,67 +417,34 @@ class StateAssemblerMarginBindingAdaptersTest {
-      getExplorationMarginEnd(),
-      getExplorationMarginBottom()
+  private fun convertToPx(dip: Float): Float {
+    val r: Resources = context.resources
```
Here & in other suites: prefer ``resources`` for the variable name. We shouldn't use single-letter variables names except in for-loops.

veena14cs replied:
Done.

BenHenning reviewed (changes requested):
Thanks @veena14cs. I have just a few more comments.

Regarding my request to make sure that the tests are failing when expected, to clarify one bit: I'm actually suggesting that you temporarily revert the RTL code in the adapters classes and then run the tests to make sure that all of the RTL-specific tests fail. Is that what you did when you encountered the findings that only one of the tests were actually failing?

veena14cs committed:
fixed lint issues

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs commented:
> Thanks @veena14cs. I have just a few more comments.
> 
> Regarding my request to make sure that the tests are failing when expected, to clarify one bit: I'm actually suggesting that you temporarily revert the RTL code in the adapters classes and then run the tests to make sure that all of the RTL-specific tests fail. Is that what you did when you encountered the findings that only one of the tests were actually failing?

None of the tests were failing when I revert the RTL code in adapter. The test case which was failing was just to show by flipping the values RTL fails, that was just for sample and I have removed this test..

But now this thing is fixed. Actually, I was making mistake while setting margins. I was setting margin and then changing the layout direction, but I had to do it other way i.e first change the layout direction and then set the margin using binding adapters. After this it will check for RTL code in adapter classes and now if if this RTL code is temporarily removed the tests fails as expected. I have added screenshots in PR description.

veena14cs committed:
fixed margin binding adapters tests

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs commented:
@BenHenning  https://github.com/oppia/oppia-android/pull/3452#discussion_r675369955 I fixed this and pushed the code. But this fails in Circle CI. It expects 6 indentations 

![Screenshot (35)](https://user-images.githubusercontent.com/16301028/126842302-7a795a2b-c538-4fdc-983b-e940e084fabd.png)

veena14cs committed:
Fixed tests

veena14cs committed:
reverted config changes

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

anandwana001 commented:
@veena14cs Let's update this PR with the latest develop, so that Bazel can build and we can check if our test are passing on Bazel or not.

anandwana001 commented on app/src/main/res/layout/test_margin_bindable_adapter_activity.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+  xmlns:app="http://schemas.android.com/apk/res-auto">
```
this is unused, we can remove it.

veena14cs replied:
Done.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+
+private const val TOLERANCE = 1e-5f
+
+/** Tests for [MarginBindableAdapters]. */
```
Should it be `MarginBindingAdapters`!!
as `MarginBindableAdapters` is nothing we have currently.

veena14cs replied:
It has to be  `MarginBindingAdapters`. Fixed.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+    )
+  }
+
+  private fun convertToPx(dip: Float): Float {
```
```suggestion
  private fun dipToPixels(dip: Float): Float {
```

anandwana001 replied:
Let's move this function to `oppia-android/utility/src/main/java/org/oppia/android/util/`. 

Also, this sounds like a good first issue too, but we can move this function and create an issue to use the function from the utility at all other places for these test files.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt:
```diff
@@ -0,0 +1,328 @@
+      return@runWithActivity textView
+    }
+    assertThat(textView.marginStart.toFloat()).isWithin(TOLERANCE).of(
+      convertToPx(24f)
```
named argument for all such functions

veena14cs replied:
Done.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,516 @@
+    ActivityScenarioRule(
+      Intent(
+        ApplicationProvider.getApplicationContext(),
+        StateAssemblerMarginBindingAdaptersTestActivity::class.java
```
I see all the newly introduced TestActivity are using the same layout file, so can we have a generalized TestActivity or is there anything specific that's why we are using different TestActivity for all?

veena14cs replied:
Nothing specific. This sounds good to me to have generalized activity, but my idea was to keep separate test file for each as there is already BinadableAdapterTestActivity, by this we may not have confusion.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,515 @@
+
+  @Config(qualifiers = "port")
+  @Test
+  fun testStateAssemblerPadding_rtlIsEnabled_port_paddingStartAndpaddingEndForRtlIsCorrect() {
```
```suggestion
  fun testStateAssemblerPadding_rtlIsEnabled_port_paddingStartAndPaddingEndForRtlIsCorrect() {
```

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,515 @@
+
+  @Config(qualifiers = "port")
+  @Test
+  fun testStateAssemblerPadding_ltrIsEnabled_port_paddingStartAndpaddingEndForLtrIsCorrect() {
```
```suggestion
  fun testStateAssemblerPadding_ltrIsEnabled_port_paddingStartAndPaddingEndForLtrIsCorrect() {
```

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,515 @@
+
+  @Config(qualifiers = "land")
+  @Test
+  fun testStateAssemblerPadding_ltrIsEnabled_landscape_paddingStartAndpaddingEndForLtrIsCorrect() {
```
```suggestion
  fun testStateAssemblerPadding_ltrIsEnabled_landscape_paddingStartAndPaddingEndForLtrIsCorrect() {
```

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,515 @@
+
+  @Config(qualifiers = "sw600dp-port")
+  @Test
+  fun testStateAssemblerPadding_ltrEnabled__port_tablet_paddingStartAndpaddingEndForLtrIsCorrect() {
```
```suggestion
  fun testStateAssemblerPadding_ltrEnabled__port_tablet_paddingStartAndPaddingEndForLtrIsCorrect() {
```

anandwana001 commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -359,11 +359,14 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/ImageRegion
 exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/ProfileChooserFragmentTestActivityPresenter.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/SplashTestActivity.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/SplashTestActivityPresenter.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/testing/StateAssemblerMarginBindingAdaptersTestActivity.kt"
```
Technically `StateAssemblerPaddingBindingAdaptersTest` is the test file for this activity. 
So, why do we need to exempt it? 

Same for the `StateAssemblerPaddingBindingAdaptersTestActivity` and `MarginBindableAdaptersTestActivity`. 

If the CI check is failing for this, We need to check why the CI is failing, is it because of file namings?

veena14cs replied:
Circle ci was failing so I had to add this file here.

Sparsh1212 replied:
The check is failing because of the file naming. Basically, if the name of the file says "ProdFile.kt", then the test file name must be "ProdFileTest.kt" in order for the check to pass.

veena14cs replied:
Thanks @Sparsh1212 fixed this.

anandwana001 replied:
any update here?

veena14cs replied:
Yes it's fixed.

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt

Co-authored-by: Akshay Nandwana <akshaynandwana001@gmail.com>

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt

Co-authored-by: Akshay Nandwana <akshaynandwana001@gmail.com>

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt

Co-authored-by: Akshay Nandwana <akshaynandwana001@gmail.com>

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt

Co-authored-by: Akshay Nandwana <akshaynandwana001@gmail.com>

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindableAdaptersTest.kt

Co-authored-by: Akshay Nandwana <akshaynandwana001@gmail.com>

veena14cs committed:
fixed changes

veena14cs committed:
Merge branch 'fix-exploration-player-texts' of https://github.com/veena14cs/oppia-android into fix-exploration-player-texts

veena14cs commented:
> @veena14cs Let's update this PR with the latest develop, so that Bazel can build and we can check if our test are passing on Bazel or not.

PR is already up-to-date with develop. Bazel tests are still failing.

veena14cs committed:
created common function in utils

veena14cs committed:
opyimized and fixed issues

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
fixed empted file issue

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Removed conversion function

anandwana001 commented:
> > @veena14cs Let's update this PR with the latest develop, so that Bazel can build and we can check if our test are passing on Bazel or not.
> 
> PR is already up-to-date with develop. Bazel tests are still failing.

The reason Bazel is not able to build here is that we haven't told Bazel about our new `*Adapters.java` files. 
Please check `BINDING_ADAPTERS` at file `app/BUILD.bazel`. 

Check the doc of both `BINDING_ADAPTERS_WITH_RESOURCE_IMPORTS` and `BINDING_ADAPTERS`. 
What we do in Bazel here is get a list of all the files which are related to view model, views, listener, annotations and binding adapters and tell Bazel to not include these files while creating the kotlin library.  `kt_android_library` is nothing but a Bazel rule which internally creates targets.

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
fixed merge conflict

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
Update BUILD.bazel

anandwana001 commented on app/BUILD.bazel:
```diff
@@ -387,6 +387,9 @@ VIEWS = [
 BINDING_ADAPTERS_WITH_RESOURCE_IMPORTS = [
     "src/main/java/org/oppia/android/app/databinding/DrawableBindingAdapters.java",
     "src/main/java/org/oppia/android/app/databinding/ImageViewBindingAdapters.java",
+    "src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java",
+    "src/main/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdapters.java",
+    "src/main/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdapters.java",
```
I think this should go into the `BINDING_ADAPTERS` only, but that's also giving the error that ViewCompat is coming from external dependency. 

Also, we need to add these new Test files to `kt_android_library` naming `test_deps`. 
Till now we had tried these things separately, lets fix them all and check if the build passes or not. 

Also, do we need any kind of dependency inside `TEST_DEPS`, is there anything currently which provides ViewCompat, 
I think the dependency should be `androidx.core:core-ktx:1.0.2`, which is there for android library `databinding_resources`, but not for the tests.

veena14cs replied:
I have updated the changes and pushed the code. Let me see if it is resolved.

FareesHussain replied:
@veena14cs you need to include the viewcompact in the [deps](https://github.com/oppia/oppia-android/blob/develop/app/BUILD.bazel#L493) for the BINDING_ADAPTERS
I think the dependency is
```
"//third_party:androidx_core_core"
```

@anandwana001 I didn't get the ktx part here, we are using java right, does java support that?

FareesHussain replied:
Ah it is already added

veena14cs replied:
@FareesHussain yes right. Did you get what is missing.?

FareesHussain replied:
I'm unable to figure it out.
@fsharpasharp, @BenHenning any idea when do we get the following error
```
app/src/main/java/org/oppia/android/app/databinding/MarginBindingAdapters.java:97: error: [strict] Using type androidx.core.view.ViewCompat from an indirect dependency (TOOL_INFO: "bazel-out/android-armeabi-v7a-fastbuild/bin/external/maven/_aar/androidx_core_core/classes_and_libs_merged.jar").
    return ViewCompat.getLayoutDirection(view) == ViewCompat.LAYOUT_DIRECTION_RTL;
``` 

I don't understand when indirect dependency mean here.

veena14cs committed:
added test file in Bazel and library in Test_deps

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
removed test files

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTestActivityTest.kt:
```diff
@@ -85,10 +84,10 @@ private const val TOLERANCE = 1e-5f
   qualifiers = "port-xxhdpi"
 )
-class StateAssemblerMarginBindingAdaptersTest {
+class StateAssemblerMarginBindingAdaptersTestActivityTest {
```
These renames don't quite make sense to me. This is testing the binding adapters, not the test activity. I suggest reverting the naming changes for this & the other test suites.

veena14cs replied:
The change was because we didn't wanted to exempt file as it was discussed with @Sparsh1212 and @anandwana001 on hangouts chat.

I will revert these changes.

Sparsh1212 replied:
As per what I can see, there is this Activity added: `StateAssemblerMarginBindingAdaptersTestActivity`.
Now, the test file check expects that we must write a test for this test activity, and the name of that file which tests this activity must be `StateAssemblerMarginBindingAdaptersTestActivityTest`.
Here, above if we are not testing the activity but the binding adapters, then the older naming was correct, as here we aren't testing the activity but the binding adapters. 
Concluding, I suggest to rename this to the earlier version just as Ben said and add a test for the test activity in a new file and please name it as: `StateAssemblerMarginBindingAdaptersTestActivityTest`

veena14cs replied:
What I would be testing in this file?

Isn't it is good to add in exempted files?

Sparsh1212 replied:
> What I would be testing in this file?
> 
> Isn't it is good to add in exempted files?

WDYT @BenHenning ?

BenHenning replied:
@Sparsh1212 the issue is that this test intends to verify the public API of the margin binding adapters (as a specific example), not the test activity. We could have a separate test suite to verify the test activity itself (meaning 2 for each pair of activity & binding adapters), but that has diminishing returns vs. the time needed to write the tests, and we plan to remove all of these test activities in the long term.

We shouldn't name something just to avoid the exemption. Tests should be named based on the thing they're testing.

Marking as resolved since the latest changes reverted the names.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTestActivityTest.kt:
```diff
@@ -117,44 +116,44 @@ class StateAssemblerMarginBindingAdaptersTest {
-    )
     assertThat(textView.marginStart.toFloat()).isWithin(TOLERANCE).of(
-      convertToPx(40f)
+      /* marginStart= */ 24f
```
Using argument comments doesn't quite make sense in this context, I think. The comment always corresponds to the argument name; ``of()`` doesn't have an argument with name ``marginStart``. Beyond that, the purpose of this comment is to provide context for literals. This assertion line has sufficient context since it's evaluating ``marginStart``. Please remove the argument comment here.

Ditto for all other assertion lines in this & all other test suites in this PR.

veena14cs replied:
Thanks for clarification. Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTestActivityTest.kt:
```diff
@@ -209,110 +207,110 @@ class StateAssemblerPaddingBindingAdaptersTest {
     }
     assertThat(textView.paddingStart.toFloat()).isWithin(TOLERANCE).of(
-      convertToPx(16f)
+      /* paddingStart= */ 12f
```
Shouldn't this be 16? It doesn't seem like RTL is working here since this code is verifying that padding start/end never changed.

veena14cs replied:
Actually, I was in wrong assumption, due to which I created confusion. Only binding attributes will flip the values as it doesn't understand the device is in RTL ,so if we have `left padding/marging 12f` and `right padding/margin 16f `. In RTL roboelectric reads from right to left i.e `start padding/marging 12f` and `end padding/marging 16f`. Hope this makes sense.

BenHenning replied:
Ah, that is actually really confusing but it now makes sense to me why these tests correctly fail per your screenshots.

I actually think we need to document this. For each of these tests, I think you need to add a comment something to the effect of:

```kotlin
// Note that the padding starts/ends below match the ones set above because, when the adapters are working correctly in RTL mode, the start/end should be exactly the start/end originally set.
```

(Needs 100-char line wrapping and might be able to be shortened slightly).

This is a confusing enough context that I think we need to be really clear about it.

veena14cs replied:
Added comments for all the testcases.

BenHenning reviewed (changes requested):
Thanks @veena14cs. Took another pass.

I'm a bit concerned that some of the RTL tests were flipped--that doesn't seem right. We should be verifying that the start/end margins are reversed, not that they're still the same.

BenHenning commented:
> @BenHenning [#3452 (comment)](https://github.com/oppia/oppia-android/pull/3452#discussion_r675369955) I fixed this and pushed the code. But this fails in Circle CI. It expects 6 indentations
> 
> ![Screenshot (35)](https://user-images.githubusercontent.com/16301028/126842302-7a795a2b-c538-4fdc-983b-e940e084fabd.png)

Ah, sorry for pushing on the +4 indentation. Let's follow the Java linter's advice here.



> > Thanks @veena14cs. I have just a few more comments.
> > Regarding my request to make sure that the tests are failing when expected, to clarify one bit: I'm actually suggesting that you temporarily revert the RTL code in the adapters classes and then run the tests to make sure that all of the RTL-specific tests fail. Is that what you did when you encountered the findings that only one of the tests were actually failing?
> 
> None of the tests were failing when I revert the RTL code in adapter. The test case which was failing was just to show by flipping the values RTL fails, that was just for sample and I have removed this test..
> 
> But now this thing is fixed. Actually, I was making mistake while setting margins. I was setting margin and then changing the layout direction, but I had to do it other way i.e first change the layout direction and then set the margin using binding adapters. After this it will check for RTL code in adapter classes and now if if this RTL code is temporarily removed the tests fails as expected. I have added screenshots in PR description.

Ah, glad you figured it out! Also, thanks for adding the screenshots to the opening comment.

veena14cs commented:
> Thanks @veena14cs. Took another pass.
> 
> I'm a bit concerned that some of the RTL tests were flipped--that doesn't seem right. We should be verifying that the start/end margins are reversed, not that they're still the same.

I have replied this on comment https://github.com/oppia/oppia-android/pull/3452#discussion_r678159711

veena14cs committed:
fixed suggested changes

Sparsh1212 reviewed (approved):
LGTM for Script CI checks as it is green and no file is added to test file exemptions.

veena14cs commented:
> > @BenHenning [#3452 (comment)](https://github.com/oppia/oppia-android/pull/3452#discussion_r675369955) I fixed this and pushed the code. But this fails in Circle CI. It expects 6 indentations
> > ![Screenshot (35)](https://user-images.githubusercontent.com/16301028/126842302-7a795a2b-c538-4fdc-983b-e940e084fabd.png)
> 
> Ah, sorry for pushing on the +4 indentation. Let's follow the Java linter's advice here.

Done, thanks.

veena14cs commented:
> LGTM for Script CI checks as it is green and no file is added to test file exemptions.

@Sparsh1212 Can you please look into this comment

Sparsh1212 commented:
> > LGTM for Script CI checks as it is green and no file is added to test file exemptions.
> 
> @Sparsh1212 Can you please look into this comment

Ah! Can you please include the link to the comment?

veena14cs commented:
> > > LGTM for Script CI checks as it is green and no file is added to test file exemptions.
> > 
> > 
> > @Sparsh1212 Can you please look into this comment
> 
> Ah! Can you please include the link to the comment?

Oh sorry here it is https://github.com/oppia/oppia-android/pull/3452#discussion_r677998296.

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -166,12 +166,15 @@
       android:name=".app.testing.ProfileChooserFragmentTestActivity"
       android:theme="@style/OppiaThemeWithoutActionBar" />
+    <activity android:name=".app.testing.StateAssemblerPaddingBindingAdaptersTestActivity" />
+    <activity android:name=".app.testing.StateAssemblerMarginBindingAdaptersTestActivity" />
```
This should come above 
```
    <activity android:name=".app.testing.StateAssemblerMarginBindingAdaptersTestActivity" />
    <activity android:name=".app.testing.StateAssemblerPaddingBindingAdaptersTestActivity" />
```

veena14cs replied:
Done.

rt4914 reviewed (approved):
Approving for code owner files. One nit change suggested.

veena14cs committed:
Update AndroidManifest.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

BenHenning reviewed (changes requested):
Thanks @veena14cs! Just had 1 comment about adding more context in the tests, but otherwise the PR LGTM. Please assign back once the final comment is resolved.

Edit: actually, you need to also fix the Bazel failures. I think you just need to add a dependency for ViewCompat in the android_library for these adapters (Java libraries don't automatically get indirect dependencies unlike Kotlin).

veena14cs committed:
Added comments

veena14cs committed:
Update BUILD.bazel

anandwana001 reviewed (changes requested):
Test file LGTM,

Please update it from the latest the develop. Need to check on how to fix the bazel issue. Fix the new checks Kdoc and accessibility 
https://github.com/oppia/oppia-android/pull/3452/checks?check_run_id=3201126942

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
fixed kdoc

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Added kdoc where required

veena14cs committed:
Added kdoc

veena14cs committed:
Update StateAssemblerPaddingBindingAdaptersTest.kt

veena14cs committed:
Update StateAssemblerPaddingBindingAdaptersTest.kt

veena14cs committed:
Update accessibility_label_exemptions.textproto

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Trigger Build

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

BenHenning commented on app/BUILD.bazel:
```diff
@@ -807,6 +810,8 @@ crashlytics_android_library(
 android_library(
     name = "crashlytics_deps",
     exports = [
+        "//third_party:androidx_core_core",
+        "//third_party:androidx_core_core-ktx",
```
This doesn't seem correct--why do we need to export these here?

veena14cs replied:
Removed this library from here.

Edit: Circle ci was failing here if I don't add `androidx_core_core` here.

<img width="384" alt="Screenshot (52)" src="https://user-images.githubusercontent.com/16301028/128197534-3fa7e8fc-9ece-4f22-ab47-9dfe34c934ac.png">

Removed `androidx_core_core-ktx` library.

BenHenning replied:
There isn't enough information from that screenshot to see which dependency is failing to build. Building it locally showed me that the only dep changes needed for this PR:
- Adding androidx_core_core to view_models
- Adding androidx_core_core to TEST_DEPS (this might not actually be needed, but prefer adding it)

None of the other changes are needed (including the new core-ktx as they aren't used).

I'm not actually 100% sure why view_models and databinding libraries need this--that's not expected. There might be some weird dependency swapping happening here by maven_install, but I can't be sure. Either way, the changes above seem reasonable.

BenHenning replied:
This isn't fully resolved yet--PTAL at my latest comment.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt:
```diff
@@ -0,0 +1,326 @@
+      setLayoutMarginEnd(textView, /* marginEnd= */ 40f)
+      return@runWithActivity textView
+    }
+    /**
+     * Note that the margin starts/ends below match the ones set above because, when the adapters are
+     * working correctly in RTL mode, the start/end should be exactly the start/end originally set.
+     */
```
This should be a standard block comment, not a KDoc:

```suggestion
    /*
     * Note that the margin starts/ends below match the ones set above because, when the adapters are
     * working correctly in RTL mode, the start/end should be exactly the start/end originally set.
     */
```

Ditto elsewhere below & in the other tests.

BenHenning reviewed (commented):
Thanks @veena14cs. LGTM except two small things--PTAL.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs committed:
fixed block comment and updated bazel

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs commented:
> Thanks @veena14cs. LGTM except two small things--PTAL.

Thanks @BenHenning I have done changes. PTAL.

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -147,10 +147,10 @@ class StateAssemblerPaddingBindingAdaptersTest {
+    /*
+    * Note that the padding starts/ends below match the ones set above because, when the adapters are
+    * working correctly in RTL mode, the start/end should be exactly the start/end originally set.
+    */
```
```suggestion
    /*
     * Note that the padding starts/ends below match the ones set above because, when the adapters are
     * working correctly in RTL mode, the start/end should be exactly the start/end originally set.
     */
```

Nit on formatting--please fix elsewhere.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs committed:
Update StateAssemblerPaddingBindingAdaptersTest.kt

veena14cs commented:
@BenHenning In that case do I need to go ahead with the changes or revert it?

veena14cs committed:
Update BUILD.bazel

BenHenning commented on app/BUILD.bazel:
```diff
@@ -804,6 +806,7 @@ crashlytics_android_library(
 android_library(
     name = "crashlytics_deps",
     exports = [
+        "//third_party:androidx_core_core",
```
Per my other comment, please remove this & the one from binding_adapters, and add it to view_models.

veena14cs replied:
Done.

BenHenning commented on app/BUILD.bazel:
```diff
@@ -553,7 +553,7 @@ kt_android_library(
         ":snap_helper",
         ":view_models",
         "//third_party:androidx_appcompat_appcompat",
-        "//third_party:androidx_core_core-ktx",
```
I think you may have misread my earlier comment. This one is still needed otherwise the build will break.

veena14cs replied:
Done.

BenHenning reviewed (changes requested):
@veena14cs I think you might not have fully followed the suggestions in my earlier comment. PTAL at the follow-ups.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

veena14cs committed:
Update BUILD.bazel

veena14cs commented:
> @veena14cs I think you might not have fully followed the suggestions in my earlier comment. PTAL at the follow-ups.

I have corrected. PTAL.

anandwana001 reviewed (approved):
LGTM

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-exploration-player-texts

BenHenning reviewed (approved):
Thanks @veena14cs. This LGTM!
